proximal/examples/test_deconv_sinv_psf.py

```diff
@@ -0,0 +1,89 @@
+# Override the default ProxImaL installation in the system
+import sys
+sys.path.append('../../')
+
+
+import proximal as px
+from proximal.lin_ops.black_box import LinOpFactory
+from scipy.signal import convolve
+
+#===========
+# SYNTHETIC DATA
+import numpy as np
+import matplotlib.pyplot as plt
+from scipy.stats import multivariate_normal
+
+# Create Synthetic Horizontal Stripes Pattern Image
+def get_raw(width=3840, height=2748):
+    im_s = np.zeros((2748, 3840))
+    stripe_width, stripe_gap, stripe_start, stripe_end = 50, 50, 500, 500
+    for y in range(0, im_s.shape[0]-(stripe_start+stripe_end), stripe_width + stripe_gap):
+        im_s[stripe_start+y : stripe_start+y + 50, :] = .8
+
+    xx, yy = np.mgrid[0:im_s.shape[0], 0:im_s.shape[1]]
+    im_ctr = (np.array(im_s.shape)/2).astype(int)
+    r = np.sqrt((xx - im_ctr[0])**2 + (yy - im_ctr[1])**2)
+    mask = np.zeros_like(im_s)
+    mask[r < 1000] = 1
+    im_s *= mask
+
+    return np.asfortranarray(im_s, dtype=np.float32), im_ctr
+
+im_s, im_ctr = get_raw()
+plt.subplot(221); plt.imshow(im_s); plt.title('Ground Truth');
+
+# Create Gaussian Kernel & Convolve
+def get_blur_kernel(im_shape, im_ctr, crop=True):
+    x, y = np.mgrid[0:im_shape[0], 0:im_shape[1]]
+    pos = np.dstack((x, y))
+    rv = multivariate_normal.pdf(pos, im_ctr, [[200, 0], [300, 500]])
+    psf2 = np.asfortranarray(rv/np.max(rv), dtype=np.float32)
+
+    if not crop:
+        return psf2
+
+    psf2 = psf2[im_ctr[0]-100:im_ctr[0]+101, im_ctr[1]-100:im_ctr[1]+101]
+
+    return psf2
+
+psf2 = get_blur_kernel(im_s.shape, im_ctr)
+plt.subplot(222); plt.imshow(psf2); plt.title('PSF zoomed in');
+
+im_sb = np.asfortranarray(convolve(im_s, psf2, mode='same', method='auto'), dtype=np.float32)
+plt.subplot(223); plt.imshow(im_sb); plt.title('Blurred Image'); plt.colorbar()
+
+#===========
+# SOLVER
+def my_factory(psf, im_shape):
+    
+    def forward(im_src, im_sim):
+        assert np.all(im_src.shape == im_sim.shape)
+        im_sim[:] = convolve(im_src, psf, mode='same', method='auto')
+
+    def adjoint(im_src, im_sim):
+        assert np.all(im_src.shape == im_sim.shape)
+        im_sim[:] = convolve(im_src, np.rot90(psf, k=2), mode='same', method='auto')
+    
+    # Compute the norm of the Gram matrix, i.e. max || convT(conv(x)) || for all images x.
+    scaling_factor = np.sum(psf)
+    return LinOpFactory(im_shape, im_shape, forward, adjoint, norm_bound=scaling_factor)
```

Comment:
>  how do I find the norm_bound in case of the more complex forward model for the Spatially Variant PSF case? There, I'm summing up the weighted convolutions, so would it be the sum over sum over all the PSFs?

Are you asking for the analytical solution of `norm_bound`, or its numerical approximation?

I am not sure, but I can direct you to internal code for reference. For analytical solution, please read:
https://github.com/comp-imaging/ProxImaL/blob/3437814f70b45669aab5a44f7da02ee2da5751db/proximal/lin_ops/conv.py#L117-L130

For numerical approximation of the norm bound, please read Secion 5.3 of the research article, followed by studying the internal code:

https://github.com/comp-imaging/ProxImaL/blob/3437814f70b45669aab5a44f7da02ee2da5751db/proximal/lin_ops/comp_graph.py#L423-L432